File: server/kcaa/kcsapi/questlist.py
#!/usr/bin/env python

import jsonobject
import model
import resource


class Quest(jsonobject.JSONSerializableObject):

    id = jsonobject.ReadonlyJSONProperty('id', value_type=int)
    """ID."""
    name = jsonobject.ReadonlyJSONProperty('name', value_type=unicode)
    """Name."""
    description = jsonobject.ReadonlyJSONProperty('description',
                                                  value_type=unicode)
    """Description."""
    category = jsonobject.ReadonlyJSONProperty('category', value_type=int)
    """Category."""
    CATEGORY_ORGANIZING = 1
    CATEGORY_EXPEDITION = 2
    CATEGORY_PRACTICE = 3
    CATEGORY_MISSION = 4
    CATEGORY_LOGISTICS = 5
    CATEGORY_SHIPYARD = 6
    CATEGORY_REBUILDING = 7
    state = jsonobject.ReadonlyJSONProperty('state', value_type=int)
    """State."""
    STATE_INACTIVE = 1
    STATE_ACTIVE = 2
    STATE_COMPLETE = 3
    progress = jsonobject.ReadonlyJSONProperty('progress', value_type=int)
    """Progress percentile."""
    bonus_type = jsonobject.ReadonlyJSONProperty('bonus_type', value_type=int)
    """Bonus type."""
    # TODO: Investigate what these values mean. Probably what kind of bonus
    # items can be obtained.
    cycle = jsonobject.ReadonlyJSONProperty('cycle', value_type=int)
    """Repop cycle."""
    CYCLE_ONCE = 1
    CYCLE_DAILY = 2
    CYCLE_WEEKLY = 3
    rewards = jsonobject.ReadonlyJSONProperty('rewards',
                                              value_type=resource.Resource)
    """Rewards."""


class QuestList(model.KCAAObject):
    """List of quests.

    This object holds a list of all the quests. Note that this object may not
    be up to date when a user hasn't accessed the quests page.
    """

    count = jsonobject.JSONProperty('count', value_type=int)
    """Number of all quests."""
    count_undertaken = jsonobject.JSONProperty('count_undertaken',
                                               value_type=int)
    """Number of quests undertaken."""
    quests = jsonobject.JSONProperty('quests', [], value_type=list,
                                     element_type=Quest)
    """Quest instances."""

    def update(self, api_name, request, response, objects, debug):
        super(QuestList, self).update(api_name, request, response, objects,
                                      debug)
        data = jsonobject.parse(response['api_data'])
        self.count = data.api_count
        self.count_undertaken = data.api_exec_count
        quests = []
        for quest_data in data.api_list:
            if quest_data == -1:
                continue
            progress = (0, 50, 80)[quest_data.api_progress_flag]
            if quest_data.api_state == Quest.STATE_COMPLETE:
                progress = 100
            quests.append(Quest(
                id=quest_data.api_no,
                name=quest_data.api_title,
                description=quest_data.api_detail,
                category=quest_data.api_category,
                state=quest_data.api_state,
                progress=progress,
                bonus_type=quest_data.api_bonus_flag,
                cycle=quest_data.api_type,
                rewards=resource.Resource(
                    fuel=quest_data.api_get_material[0],
                    ammo=quest_data.api_get_material[1],
                    steel=quest_data.api_get_material[2],
                    bauxite=quest_data.api_get_material[3])))
        quests.sort(lambda x, y: x.id - y.id)
        self.quests = model.merge_list(self.quests, quests)

File: client/web/domain/questlist.dart
part of kcaa;

class Quest {
  int id;
  String name;
  String description;
  String category;
  String state;
  int fuel, ammo, steel, bauxite;
  int progress;
  String cycle;

  static final Map<int, String> CATEGORY_MAP = <int, String>{
    1: "編成",
    2: "出撃",
    3: "演習",
    4: "遠征",
    5: "補給",
    6: "工廠",
    7: "改装",
  };
  static final Map<int, String> STATE_MAP = <int, String>{
    1: "",
    2: "active",
    3: "complete",
  };
  static final Map<int, String> CYCLE_MAP = <int, String>{
    1: "一回",
    2: "日毎",
    3: "週毎",
  };

  Quest(this.id, this.name, this.description, int category, int state,
      Map<String, int> rewards, int progress, int cycle)
      : category = CATEGORY_MAP[category],
        state = STATE_MAP[state],
        fuel = rewards["fuel"],
        ammo = rewards["ammo"],
        steel = rewards["steel"],
        bauxite = rewards["bauxite"],
        progress = progress,
        cycle = CYCLE_MAP[cycle] {}
}

void handleQuestList(Assistant assistant) {
  assistant.getObject("QuestList", false).then((Map<String, dynamic> data) {
    assistant.numQuests = data["count"];
    assistant.numQuestsUndertaken = data["count_undertaken"];
    assistant.quests.clear();
    for (var quest in data["quests"]) {
      assistant.quests.add(new Quest(quest["id"], quest["name"],
          quest["description"], quest["category"], quest["state"],
          quest["rewards"], quest["progress"], quest["cycle"]));
    }
  });
}